File: backend/app/schemas/sandbox.py
from __future__ import annotations

from datetime import datetime
from typing import Any, Dict, List, Optional
from uuid import UUID

from pydantic import BaseModel, Field

from app.schemas.workspace import WorkspaceAgentPlacement


class SandboxWorkspaceCreate(BaseModel):
    name: str = Field(..., min_length=1, max_length=255)
    description: Optional[str] = None
    agent_placements: List[WorkspaceAgentPlacement] = Field(default_factory=list)


class SandboxEventCreate(BaseModel):
    scenario_name: str = Field(..., min_length=1, max_length=255)
    domain: str = Field(..., min_length=1, max_length=100)
    intent: str = Field(..., min_length=1, max_length=100)
    message: str = Field(..., min_length=1, max_length=4000)
    priority: str = Field(default="medium", pattern="^(low|medium|high|critical)$")
    tags: List[str] = Field(default_factory=list)


class SandboxDecisionRead(BaseModel):
    sequence: int
    agent_id: UUID
    agent_name: str
    action: str
    reason: str
    status: str


class SandboxIsolationRead(BaseModel):
    mode: str = "sandbox_only"
    message_headers_written: int = 0
    messages_written: int = 0
    interactions_written: int = 0


class SandboxRunRead(BaseModel):
    run_id: UUID
    workspace_id: UUID
    created_by: UUID
    scenario_name: str
    status: str
    event: Dict[str, Any]
    decision_log: List[SandboxDecisionRead]
    routed_agent_ids: List[UUID]
    production_write_count: int
    isolation: SandboxIsolationRead = Field(default_factory=SandboxIsolationRead)
    created_at: datetime
    completed_at: Optional[datetime] = None

    model_config = {"from_attributes": True}

File: frontend/src/api/sandbox.ts
import client from './client';
import type { Workspace } from './workspaces';


export interface SandboxDecision {
  sequence: number;
  agent_id: string;
  agent_name: string;
  action: 'consume' | 'handoff' | 'skip';
  reason: string;
  status: 'SIMULATED' | 'SKIPPED';
}

export interface SandboxRun {
  run_id: string;
  workspace_id: string;
  created_by: string;
  scenario_name: string;
  status: 'COMPLETED' | 'FAILED';
  event: {
    domain: string;
    intent: string;
    message: string;
    priority: string;
    tags: string[];
  };
  decision_log: SandboxDecision[];
  routed_agent_ids: string[];
  production_write_count: number;
  isolation: {
    mode: 'sandbox_only';
    message_headers_written: number;
    messages_written: number;
    interactions_written: number;
  };
  created_at: string;
  completed_at?: string | null;
}

export const sandboxApi = {
  listWorkspaces: async (): Promise<Workspace[]> => {
    const response = await client.get<Workspace[]>('/sandbox/workspaces');
    return response.data;
  },

  createWorkspace: async (payload: {
    name: string;
    description?: string;
    agent_placements: Array<{ agent_id: string; quantity: number }>;
  }): Promise<Workspace> => {
    const response = await client.post<Workspace>('/sandbox/workspaces', payload);
    return response.data;
  },

  listRuns: async (workspaceId: string): Promise<SandboxRun[]> => {
    const response = await client.get<SandboxRun[]>(`/sandbox/workspaces/${workspaceId}/runs`);
    return response.data;
  },

  run: async (
    workspaceId: string,
    payload: {
      scenario_name: string;
      domain: string;
      intent: string;
      message: string;
      priority: 'low' | 'medium' | 'high' | 'critical';
      tags: string[];
    }
  ): Promise<SandboxRun> => {
    const response = await client.post<SandboxRun>(`/sandbox/workspaces/${workspaceId}/runs`, payload);
    return response.data;
  },
};
